### PR TITLE
feat: allowing for sorting and filtering of the enterprise group learner endpoints

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
https://2u-internal.atlassian.net/browse/ENT-8503
**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/openedx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/openedx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/openedx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/openedx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/openedx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: tests/test_enterprise/api/test_views.py [hunk 1]
```diff
@@ -7400,6 +7401,101 @@ def test_list_learner_statuses(self):
...
+        # Test sorting by the three sortable values
+        for sort_by_value in ['recentAction', 'memberStatus', 'memberDetails']:
+            url = settings.TEST_SERVER + reverse(
+                'enterprise-group-learners',
+                kwargs={'group_uuid': self.group_1.uuid},
+            ) + f"?q={sort_by_value}"
+            response = self.client.get(url)
+            results = response.json().get('results')
+            returned_sorted_values = [
+                value.get(re.sub(r'(?<!^)(?=[A-Z])', '_', sort_by_value).lower()) for value in results
```
Comment (marlonkeating) [inline tests/test_enterprise/api/test_views.py:7476]: There's already a helper method in this repo: https://github.com/openedx/edx-enterprise/blob/0e78b47c2d873c932e46826e031c357a5d0f9e76/enterprise/management/commands/manufacture_data.py#L56

File: enterprise/models.py [hunk 2]
```diff
@@ -4259,11 +4260,24 @@ class Meta:
...
+            sort_by (optional): specify how the list of returned members should be ordered. Supported sorting values
+                                are `memberDetails`, `memberStatus`, and `recentAction`. Ordering can be reversed by
+                                supplying a `-` at the beginning of the sorting value ie `-memberStatus`.
+        """
+        if q:
+            # pylint: disable=unsupported-binary-operation
+            # unfortunately, user is a property not a field on the ecu model. In order to filter ecu's by name or email
+            # means we have to filter users by name/email and then apply fetched user ID's to the ecu table
+            user_ids = User.objects.filter(
+                Q(email__contains=q) | Q(first_name__contains=q) | Q(last_name__contains=q)
```
Comment (marlonkeating) [inline enterprise/models.py:4279]: Should we be using `__icontains` for case-insensitive searching here?

**Review by marlonkeating — CHANGES_REQUESTED** [review]

File: enterprise/models.py [hunk 3]
```diff
@@ -4344,6 +4377,37 @@ def membership_user(self):
...
+            return "accepted"
+        return "pending"
+
+    @property
+    def recent_action(self):
+        """
+        Return the timestamp of the most recent action relating to the membership
+        """
+        if self.is_removed:
+            return self.modified
```
Comment (kiram15) [inline enterprise/models.py:4406]: Can something not be modified after it's removed?

File: tests/test_enterprise/api/test_views.py [hunk 4]
```diff
@@ -7400,6 +7401,100 @@ def test_list_learner_statuses(self):
         statuses = [result.get('member_status') for result in response.json().get('results')]
         assert statuses.sort() == ['accepted', 'pending'].sort()
 
+    def test_list_learners_filtered(self):
```
Comment (kiram15) [inline tests/test_enterprise/api/test_views.py:7404]: nit: not blocking but might want to include a test with the page being passed in?

**Review by kiram15 — APPROVED** [review]
Looks good! Just one question and one nit

File: enterprise/models.py (hunk 3 above)
Comment (alex-sheehan-edx) [reply enterprise/models.py:4406]: potentially manually, but not programmatically AFAIK

File: tests/test_enterprise/api/test_views.py (hunk 4 above)
Comment (alex-sheehan-edx) [reply tests/test_enterprise/api/test_views.py:7404]: I believe @katrinan029 already has a pagination test! https://github.com/openedx/edx-enterprise/blob/master/tests/test_enterprise/api/test_views.py#L7437-L7440

File: tests/test_enterprise/api/test_views.py [hunk 5]
```diff
@@ -7400,6 +7401,103 @@ def test_list_learner_statuses(self):
...
+            if sort_by_value == 'memberDetails':
+                assert returned_sorted_values == sorted(returned_sorted_values, key=lambda t: t.get('user_email'))
+            else:
+                assert returned_sorted_values == sorted(returned_sorted_values)
+
+        # Test sorting in reverse order
+        url = settings.TEST_SERVER + reverse(
+            'enterprise-group-learners',
+            kwargs={'group_uuid': self.group_1.uuid},
+        ) + "?q=-memberDetails"
```
Comment (marlonkeating) [inline tests/test_enterprise/api/test_views.py:7489]:
```suggestion
        ) + "?sort_by=-memberDetails"
```

File: tests/test_enterprise/api/test_views.py [hunk 6]
```diff
@@ -7400,6 +7401,103 @@ def test_list_learner_statuses(self):
...
+        """
+        # Test sorting by the three sortable values
+        for sort_by_value in ['recentAction', 'memberStatus', 'memberDetails']:
+            url = settings.TEST_SERVER + reverse(
+                'enterprise-group-learners',
+                kwargs={'group_uuid': self.group_1.uuid},
+            ) + f"?q={sort_by_value}"
+            response = self.client.get(url)
+            results = response.json().get('results')
+            returned_sorted_values = [value.get(convert_to_pascal(sort_by_value)) for value in results]
```
Comment (marlonkeating) [inline tests/test_enterprise/api/test_views.py:7479]: Does the serializer export in snake case or pascal case?

**Review by marlonkeating — CHANGES_REQUESTED** [review]

File: tests/test_enterprise/api/test_views.py (hunk 5 above)
Comment (alex-sheehan-edx) [reply tests/test_enterprise/api/test_views.py]: good catch!

File: tests/test_enterprise/api/test_views.py (hunk 6 above)
Comment (alex-sheehan-edx) [reply tests/test_enterprise/api/test_views.py:7479]:
```
    class Meta:
        model = models.EnterpriseGroupMembership
        fields = (
            'learner_id',
            'pending_learner_id',
            'enterprise_group_membership_uuid',
            'member_details',
            'recent_action',
            'member_status',
        )
```

Backend behavior is to surface data with snake case. I guess the follow up question is whether the endpoint should accept values in pascal or snake case.... My original thinking was that the frontend would want to deal in pascal and to support it but I'm ok enforcing the rule on consumers ¯\_(ツ)_/¯

Comment (alex-sheehan-edx) [reply tests/test_enterprise/api/test_views.py:7479]: disregard my previous comment, I see what you mean! The test wasn't defensive enough against my craziness.

**Review by marlonkeating — APPROVED** [review]

**Review by iloveagent57 — CHANGES_REQUESTED** [review]
We've talked offline about this - there's a couple of problematic things in here that shouldn't be merged, namely passing in an arbitrary query and the poor performance of that User query.  Requesting changes to block merging.

File: enterprise/api/v1/views/enterprise_group.py [hunk 7]
```diff
@@ -117,11 +124,12 @@ def get_learners(self, *args, **kwargs):
             }
 
         """
-
+        q = self.request.query_params.get('q')
```
Comment (iloveagent57) [inline enterprise/api/v1/views/enterprise_group.py:127]: We should really do some validation of these query params before sending them off to the ORM. You could make a serializer class that you pass `self.request.query_params` into and call `is_valid()` on.

File: enterprise/models.py [hunk 8]
```diff
@@ -4259,11 +4260,34 @@ class Meta:
...
+            var_q = f"%{q}%"
+            sql_string = """
+                select entcu.id from enterprise_enterprisecustomeruser entcu
+                join auth_user au on entcu.user_id = au.id
+                where entcu.enterprise_customer_id = %s and au.email like %s;
+            """
+            # Raw sql is picky about uuid format
+            customer_id = str(self.enterprise_customer.pk).replace("-", "")
+            ecus = EnterpriseCustomerUser.objects.raw(sql_string, (customer_id, var_q))
+            ecus_ids_list = [ecu.id for ecu in ecus]
```
Comment (iloveagent57) [inline enterprise/models.py:4290]: This method is getting long and has lots of conditionals that make it hard to follow.  You could break this piece off into a helper like `def _get_filtered_ecu_ids(self, user_query=None)`.

File: enterprise/models.py [hunk 9]
```diff
@@ -4259,11 +4260,34 @@ class Meta:
...
         if self.applies_to_all_contexts:
             members = []
```
Comment (iloveagent57) [inline enterprise/models.py:4292]: Also, this whole chunk could be something like `def _get_implicit_group_members(self, user_query=None, ecu_ids=None)`.

File: enterprise/models.py [hunk 10]
```diff
@@ -4285,9 +4313,24 @@ def get_all_learners(self):
...
                     group=self,
                 ))
-            return members
         else:
-            return self.members.filter(is_removed=False)
+            members = self.members.filter(is_removed=False)
+            if q:
+                ecu_filter = Q(enterprise_customer_user__id__in=ecus_ids_list)
+                pecu_filter = Q(pending_enterprise_customer_user__user_email__icontains=q)
+                members = members.filter(ecu_filter | pecu_filter)
```
Comment (iloveagent57) [inline enterprise/models.py:4321]: You should try to check how many SQL statements this generates. I think you could add a select related to your `self.members.filter(...)` to get it down to only a single query.

File: enterprise/models.py [hunk 11]
```diff
@@ -4344,6 +4387,37 @@ def membership_user(self):
...
+    @property
+    def member_status(self):
+        """
+        Return the status of the membership
+        """
+        if self.is_removed:
+            return "removed"
+        if self.enterprise_customer_user:
+            return "accepted"
+        return "pending"
```
Comment (iloveagent57) [inline enterprise/models.py:4408]: These should be constants.  Also, this is a good indication that maybe you want to add a field called `status` to your membership model.

**Review by iloveagent57 — CHANGES_REQUESTED** [review]

File: enterprise/api/v1/views/enterprise_group.py [hunk 12]
```diff
@@ -117,11 +124,29 @@ def get_learners(self, *args, **kwargs):
...
+            is_reversed = sort_by[0] == '-'
+            sort_by = utils.convert_to_snake(sort_by.lstrip('-'))
+            query_params['sort_by'] = sort_by
+
+        param_serializers = serializers.EnterpriseGroupLearnersRequestQuerySerializer(
+            data=query_params
+        )
+        param_serializers.is_valid()
+
+        if not param_serializers.is_valid():
```
Comment (iloveagent57) [inline enterprise/api/v1/views/enterprise_group.py:139]: you don't need this call twice

File: enterprise/models.py [hunk 13]
```diff
@@ -4259,35 +4264,107 @@ class Meta:
...
+            ))
+        return members
+
+    def _get_explicit_group_members(self, user_query=None, fetch_removed=False):
+        """
+        Fetch explicitly defined members of a group, indicated by an existing membership record
+        """
+        members = self.members.filter(is_removed=fetch_removed).select_related(
+            'enterprise_customer_user', 'pending_enterprise_customer_user'
+        )
```
Comment (iloveagent57) [inline enterprise/models.py:4332]: Just calling this out: if `fetch_removed` is True, these members will be _only_ the members that have been removed from the group.

File: enterprise/models.py [hunk 14]
```diff
@@ -4259,35 +4264,107 @@ class Meta:
...
+        if self.applies_to_all_contexts:
+            members = self._get_implicit_group_members(user_query)
         else:
-            return self.members.filter(is_removed=False)
+            members = self._get_explicit_group_members(user_query, fetch_removed)
+        if sort_by:
+            desc_order = False
+            if sort_by[0] == '-':
+                desc_order = True
+                sort_by = sort_by[1:]
```
Comment (iloveagent57) [inline enterprise/models.py:4360]: You're already doing a bunch of computation around this `-` at the view layer, why don't you add another parameter to this function called `reversed` and have `sort_by` be only the name of the key to sort by, without the `-`?

File: enterprise/api/v1/views/enterprise_group.py [hunk 15]
```diff
@@ -117,11 +124,29 @@ def get_learners(self, *args, **kwargs):
             }
 
         """
+        query_params = self.request.query_params.copy()
+        is_reversed = False
+        if sort_by := query_params.get('sort_by'):
+            is_reversed = sort_by[0] == '-'
+            sort_by = utils.convert_to_snake(sort_by.lstrip('-'))
+            query_params['sort_by'] = sort_by
```
Comment (iloveagent57) [inline enterprise/api/v1/views/enterprise_group.py:132]: wouldn't things be much simpler if you accepted an optional, boolean `reversed` query parameter?

File: tests/test_enterprise/api/test_views.py [hunk 16]
```diff
@@ -7397,9 +7398,132 @@ def test_list_learner_statuses(self):
...
+        long_query_string = "foo".join('bar' for _ in range(320))
+        url = settings.TEST_SERVER + reverse(
+            'enterprise-group-learners',
+            kwargs={'group_uuid': self.group_1.uuid},
+        ) + f"?user_query={long_query_string}"
+        response = self.client.get(url)
+        assert response.status_code == 400
+        assert response.data.get('user_query')
+
+    def test_list_learners_filtered(self):
```
Comment (iloveagent57) [inline tests/test_enterprise/api/test_views.py:7426]: It'd be good to have an additional test in here to protect against any future changes to de-parameterize the raw sql.  Like pass in some Little Bobby Tables type of user query.
https://xkcd.com/327/

File: enterprise/models.py (hunk 14 above)
Comment (alex-sheehan-edx) [reply enterprise/models.py:4360]: hmmm not a bad idea

File: tests/test_enterprise/api/test_views.py (hunk 16 above)
Comment (alex-sheehan-edx) [reply tests/test_enterprise/api/test_views.py:7425]: done!

**Review by iloveagent57 — APPROVED** [review]
